File: web/src/services/api/tags.ts
import type { UUID } from "@/types/primitive";
import type { ListResponse } from "@/types/pagination";
import { http } from "./client";
import { ENDPOINTS } from "./endpoints";

export interface Tag {
  id: UUID;
  name: string;
  entity_type: string;
  category: string;
  description?: string | null;
  color?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TagCreate {
  name: string;
  entity_type?: string;
  category?: string;
  description?: string;
  color?: string;
}

export interface TagUpdate {
  name?: string;
  entity_type?: string;
  category?: string;
  description?: string;
  color?: string;
}

export interface TagUsageStats {
  tag_id: UUID;
  tag_name: string;
  entity_type: string;
  category: string;
  usage_by_entity_type: Record<string, number>;
  total_usage: number;
}

export interface TagCategoryOption {
  value: string;
  label: string;
  entity_type?: string | null;
}

export interface TagCategoryCreate {
  label: string;
  value?: string;
}

export interface TagCategoryUpdate {
  label: string;
}

export interface TagBulkUpdateRequest {
  ids: UUID[];
  category: string;
}

export interface TagBulkUpdateResponse {
  updated_count: number;
  failed_ids: UUID[];
  errors: string[];
  updated_tags: Tag[];
}

export type TagListFieldsMode = "selector" | "full";

interface TagListMeta {
  entity_type?: string | null;
  category?: string | null;
  fields?: TagListFieldsMode | null;
}

export type TagListResponse = ListResponse<Tag, TagListMeta>;

export const tagsApi = {
  getAll: (params?: {
    entity_type?: string;
    category?: string;
    page?: number;
    size?: number;
    fields?: TagListFieldsMode;
  }): Promise<TagListResponse> => http.get<TagListResponse>(ENDPOINTS.TAGS.BASE, params),
  getEntityTypes: (): Promise<string[]> =>
    http.get<string[]>(ENDPOINTS.TAGS.ENTITY_TYPES),
  getCategories: (entityType: string): Promise<TagCategoryOption[]> =>
    http.get<TagCategoryOption[]>(ENDPOINTS.TAGS.CATEGORIES, {
      entity_type: entityType,
    }),
  createCategory: (
    payload: TagCategoryCreate,
    entityType: string,
  ): Promise<TagCategoryOption> =>
    http.post<TagCategoryOption>(ENDPOINTS.TAGS.CATEGORIES, payload, {
      entity_type: entityType,
    }),
  renameCategory: (
    value: string,
    payload: TagCategoryUpdate,
    entityType: string,
  ): Promise<TagCategoryOption> =>
    http.patch<TagCategoryOption>(ENDPOINTS.TAGS.CATEGORY_BY_VALUE(value), payload, {
      entity_type: entityType,
    }),
  create: (tag: TagCreate): Promise<Tag> =>
    http.post<Tag>(ENDPOINTS.TAGS.BASE, tag),
  getById: (id: UUID): Promise<Tag> => http.get<Tag>(ENDPOINTS.TAGS.BY_ID(id)),
  update: (id: UUID, tag: TagUpdate): Promise<Tag> =>
    http.patch<Tag>(ENDPOINTS.TAGS.BY_ID(id), tag),
  bulkUpdateCategories: (
    payload: TagBulkUpdateRequest,
  ): Promise<TagBulkUpdateResponse> =>
    http.patch<TagBulkUpdateResponse>(ENDPOINTS.TAGS.BATCH_UPDATE, payload),
  delete: (id: UUID): Promise<void> => http.delete<void>(ENDPOINTS.TAGS.BY_ID(id)),
  getUsage: (id: UUID): Promise<TagUsageStats> =>
    http.get<TagUsageStats>(ENDPOINTS.TAGS.USAGE(id)),
  getStatsBatch: (entityType: string) =>
    http.get<{
      entity_type: string;
      tag_stats: Array<{ id: UUID; name?: string; usage_count: number }>;
      total_tags: number;
    }>(ENDPOINTS.STATS.TAGS_USAGE(entityType)),
};

File: src/lifeos_web/routers/tags.py
"""Tag endpoints for the local Web UI."""

from __future__ import annotations

import math
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from lifeos_cli.db.services import tags as tag_services
from lifeos_cli.db.services.tags import VALID_TAG_ENTITY_TYPES
from lifeos_web.deps import get_db_session
from lifeos_web.schemas import (
    ListResponse,
    Pagination,
    TagBulkCategoryUpdate,
    TagCategoryCreate,
    TagCategoryUpdate,
    TagCreate,
    TagUpdate,
)
from lifeos_web.serialization import to_jsonable, to_jsonable_dict

router = APIRouter(prefix="/tags", tags=["tags"])
SessionDep = Annotated[AsyncSession, Depends(get_db_session)]

DEFAULT_TAG_CATEGORIES: dict[str, tuple[str, ...]] = {
    "note": ("general", "topic"),
    "person": ("general", "location", "relation", "profession", "team"),
}


def _category_label(value: str) -> str:
    """Return the fallback label for a category value."""
    return value.replace("_", " ").title() if value else "General"


def _category_option(value: str, entity_type: str | None) -> dict[str, object]:
    normalized = tag_services.normalize_tag_category(value)
    return {
        "value": normalized,
        "label": _category_label(normalized),
        "entity_type": entity_type,
    }


def _normalize_category_value(payload: TagCategoryCreate | TagCategoryUpdate) -> str:
    raw = getattr(payload, "value", None) or payload.label
    return tag_services.normalize_tag_category(raw.replace(" ", "_"))


@router.get("/", response_model=ListResponse)
async def list_tags(
    session: SessionDep,
    page: int = Query(1, ge=1),
    size: int = Query(100, ge=1, le=1000),
    entity_type: str | None = None,
    category: str | None = None,
) -> ListResponse:
    """List LifeOS tags for selectors and tag management."""
    offset = (page - 1) * size
    try:
        rows = await tag_services.list_tags(
            session,
            entity_type=entity_type,
            category=category,
            limit=size,
            offset=offset,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    items = [to_jsonable(row) for row in rows]
    return ListResponse(
        items=items,
        pagination=Pagination(
            page=page,
            size=size,
            total=len(items),
            pages=math.ceil(len(items) / size) if size else 0,
        ),
        meta={"entity_type": entity_type, "category": category},
    )


@router.get("/entity-types/")
async def list_tag_entity_types() -> list[str]:
    """Return supported LifeOS tag entity types."""
    return sorted(VALID_TAG_ENTITY_TYPES)


@router.get("/categories/")
async def list_tag_categories(
    session: SessionDep,
    entity_type: str | None = None,
) -> list[dict[str, object]]:
    """Return built-in and persisted tag categories."""
    try:
        normalized_entity_type = (
            tag_services.validate_tag_entity_type(entity_type) if entity_type else None
        )
        categories = set(
            await tag_services.list_tag_categories(
                session,
                entity_type=normalized_entity_type,
            )
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    categories.update(DEFAULT_TAG_CATEGORIES.get(normalized_entity_type or "", ("general",)))
    return [_category_option(category, normalized_entity_type) for category in sorted(categories)]


@router.post("/categories/")
async def create_tag_category(
    payload: TagCategoryCreate,
    entity_type: str | None = None,
) -> dict[str, object]:
    """Return a normalized tag category option for the scoped entity type."""
    try:
        normalized_entity_type = (
            tag_services.validate_tag_entity_type(entity_type) if entity_type else None
        )
        return _category_option(_normalize_category_value(payload), normalized_entity_type)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.patch("/categories/{category}")
async def rename_tag_category(
    category: str,
    payload: TagCategoryUpdate,
    session: SessionDep,
    entity_type: str | None = None,
) -> dict[str, object]:
    """Rename a tag category by moving its active tags."""
    if not entity_type:
        raise HTTPException(status_code=400, detail="Tag category rename requires entity_type")
    try:
        normalized_entity_type = tag_services.validate_tag_entity_type(entity_type)
        next_category = _normalize_category_value(payload)
        await tag_services.rename_tag_category(
            session,
            entity_type=normalized_entity_type,
            category=category,
            new_category=next_category,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _category_option(next_category, normalized_entity_type)


@router.patch("/batch-update")
async def bulk_update_tag_categories(
    payload: TagBulkCategoryUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Move selected tags to another category."""
    try:
        updated_tags, failed_ids, errors = await tag_services.bulk_update_tag_categories(
            session,
            tag_ids=payload.ids,
            category=payload.category,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return {
        "updated_count": len(updated_tags),
        "failed_ids": [str(tag_id) for tag_id in failed_ids],
        "errors": errors,
        "updated_tags": [to_jsonable(tag) for tag in updated_tags],
    }


@router.post("/")
async def create_tag(payload: TagCreate, session: SessionDep) -> dict[str, object]:
    """Create a LifeOS tag."""
    try:
        tag = await tag_services.create_tag(
            session,
            name=payload.name,
            entity_type=payload.entity_type,
            category=payload.category,
            description=payload.description,
            color=payload.color,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return to_jsonable_dict(tag)


@router.get("/{tag_id}")
async def get_tag(tag_id: UUID, session: SessionDep) -> dict[str, object]:
    """Return one LifeOS tag."""
    tag = await tag_services.get_tag(session, tag_id=tag_id)
    if tag is None:
        raise HTTPException(status_code=404, detail=f"Tag {tag_id} was not found")
    return to_jsonable_dict(tag)


@router.get("/{tag_id}/usage")
async def get_tag_usage(tag_id: UUID, session: SessionDep) -> dict[str, object]:
    """Return active tagged-record count for one tag."""
    tag = await tag_services.get_tag(session, tag_id=tag_id)
    if tag is None:
        raise HTTPException(status_code=404, detail=f"Tag {tag_id} was not found")
    usage_count = await tag_services.count_tag_usage(session, tag_id=tag_id)
    return {
        "tag_id": str(tag.id),
        "tag_name": tag.name,
        "entity_type": tag.entity_type,
        "category": tag.category,
        "usage_by_entity_type": {tag.entity_type: usage_count},
        "total_usage": usage_count,
    }


@router.patch("/{tag_id}")
async def update_tag(
    tag_id: UUID,
    payload: TagUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update one LifeOS tag."""
    fields = payload.model_fields_set
    try:
        tag = await tag_services.update_tag(
            session,
            tag_id=tag_id,
            name=payload.name,
            entity_type=payload.entity_type,
            category=payload.category,
            description=payload.description,
            clear_description="description" in fields and payload.description is None,
            color=payload.color,
            clear_color="color" in fields and payload.color is None,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return to_jsonable_dict(tag)


@router.delete("/{tag_id}", status_code=204)
async def delete_tag(tag_id: UUID, session: SessionDep) -> None:
    """Soft-delete a LifeOS tag."""
    try:
        await tag_services.delete_tag(session, tag_id=tag_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
